<a href="https://colab.research.google.com/github/Camgamez/AlgorithmsUN2024II/blob/main/Lab3/CamiloGamez_Lab3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
!cat /proc/cpuinfo

processor	: 0
vendor_id	: GenuineIntel
cpu family	: 6
model		: 79
model name	: Intel(R) Xeon(R) CPU @ 2.20GHz
stepping	: 0
microcode	: 0xffffffff
cpu MHz		: 2199.998
cache size	: 56320 KB
physical id	: 0
siblings	: 2
core id		: 0
cpu cores	: 1
apicid		: 0
initial apicid	: 0
fpu		: yes
fpu_exception	: yes
cpuid level	: 13
wp		: yes
flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single ssbd ibrs ibpb stibp fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm rdseed adx smap xsaveopt arat md_clear arch_capabilities
bugs		: cpu_meltdown spectre_v1 spectre_v2 spec_store_bypass l1tf mds swapgs taa mmio_stale_data retbleed bhi
bogomips	: 4399.99
clflush size	: 64
cache_alignment	: 64
ad

# I Solving the Sum of Two Digits Programming Challenge

## C++

In [2]:
%%writefile APlusB.cpp
#include <iostream>

int sum_of_two_digits(int first_digit, int second_digit) {
    return first_digit + second_digit;
}

int main() {
    int a = 0;
    int b = 0;
    std::cin >> a;
    std::cin >> b;
    std::cout << sum_of_two_digits(a, b);
    return 0;
}

Writing APlusB.cpp


In [3]:
!g++ APlusB.cpp -o APlusB -std=c++11
!./APlusB

5 9
14

### Test input output files

In [4]:
%%writefile input1.txt
4 8

Writing input1.txt


In [5]:
%%writefile output1.txt
12

Writing output1.txt


In [6]:
%%writefile input2.txt
0 9

Writing input2.txt


In [7]:
%%writefile output2.txt
9

Writing output2.txt


In [8]:
!./APlusB < input1.txt

12

In [9]:
!./APlusB < input2.txt

9

### Comparing files with grep

https://askubuntu.com/questions/546796/comparing-two-text-files

In [10]:
!./APlusB < input1.txt> ctest1.txt
!grep -xvFf output1.txt ctest1.txt

In [11]:
!./APlusB < input2.txt > ctest2.txt
!grep -xvFf output2.txt ctest2.txt

## Python

In [12]:
%%writefile sum.py
# python3

def sum_of_two_digits(first_digit, second_digit):
    return first_digit + second_digit

if __name__ == '__main__':
    a, b = map(int, input().split())
    print(sum_of_two_digits(a, b))

Writing sum.py


In [13]:
!python sum.py

5 9
14


In [14]:
!python sum.py  < input1.txt> ptest1.txt
!grep -xvFf output1.txt ptest1.txt

In [15]:
!python sum.py  < input2.txt> ptest2.txt
!grep -xvFf output2.txt ptest2.txt

# II Solving The Maximum Pairwise Product Programming Challenge in C++

## C++

$$ T(n) < 6 n^2 = O(n^2)$$

In [16]:
%%writefile max_pairwise_product_full.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int MaxPairwiseProduct(const std::vector<int>& numbers) {
    int max_product = 0;
    int aux_product = 0;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = 0; second < n; ++second) {
          if ( first != second ){
                aux_product = numbers[first] * numbers[second];
                if (aux_product > max_product)  max_product = aux_product;
          }
        }
    }

    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}

Writing max_pairwise_product_full.cpp


#### Time Complexity
$$\left. \begin{array}{llc}
 \text{INSTRUCTION} & \text{TYPE}& \text{# EXECUTIONS}\\
\text{first != second} &  \text{comparisons} &n^2 \\
\text{numbers[first] * numbers[second]}& \ \text{ products}  &n^2 - n \\
\text{max_product = aux_product} &  \text{assigments} & [1, \frac{n^2 - n}{2})  
\end{array}\right\}  =  O(n^2)$$

In [17]:
!g++ max_pairwise_product_full.cpp -o max_pairwise_product_full
!./max_pairwise_product_full

5 14
9 5 4 8 9 5
126


### Test 1: Small numbers

In [18]:
%%writefile input0.txt
10
0 0 0 0 0 0 0 0 0 0

Writing input0.txt


In [19]:
%%writefile output0.txt
0

Writing output0.txt


In [20]:
!./max_pairwise_product_full< input0.txt > ctest0.txt
!grep -xvFf output0.txt ctest0.txt

In [21]:
%%writefile input1.txt
10
1 2 3 4 5 6 7 8 9 10

Overwriting input1.txt


In [22]:
!./max_pairwise_product < input1.txt

/bin/bash: line 1: ./max_pairwise_product: No such file or directory


In [23]:
%%writefile output1.txt
90

Overwriting output1.txt


In [24]:
!./max_pairwise_product_full< input1.txt > ctest1.txt
!grep -xvFf output1.txt ctest1.txt

In [25]:
%%writefile input2.txt
10
10 9 8 7 6 5 4 3 2 1

Overwriting input2.txt


In [26]:
%%writefile output2.txt
90

Overwriting output2.txt


In [27]:
!./max_pairwise_product_full< input2.txt > ctest2.txt
!grep -xvFf output2.txt ctest2.txt

### Test 2: largest input numbers 200000 200000 to check possible intger overflow

In [28]:
%%writefile input3.txt
6
1 3 200000 9 200000 23

Writing input3.txt


In [29]:
200000*200000

40000000000

In [30]:
%%writefile output3.txt
40000000000

Writing output3.txt


In [31]:
!./max_pairwise_product_full < input3.txt

1345294336


In [32]:
!./max_pairwise_product_full < input3.txt > ctest3.txt
!grep -xvFf output3.txt ctest3.txt

1345294336


In [ ]:
 %%writefile run_all_test.sh
#!/bin/bash
g++ max_pairwise_product_full.cpp -o max_pairwise_product_full
for i in {0..3}
do
    ./max_pairwise_product_full < input$i.txt > ctest$i.txt
    grep -xvFf output$i.txt ctest$i.txt
done

Writing run_all_test.sh


In [33]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

chmod: cannot access 'run_all_test.sh': No such file or directory
/bin/bash: line 1: ./run_all_test.sh: No such file or directory


### Changing the variables to long long

In [34]:
%%writefile max_pairwise_product_long_full.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {
    long long max_product = 0LL;
	  long long aux_product = 0LL;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = 0; second < n; ++second) {
          if ( first != second){
                aux_product = (long long) numbers[first] * numbers[second];
                if (aux_product > max_product)  max_product = aux_product;
          }
        }
    }


    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}


Writing max_pairwise_product_long_full.cpp


In [35]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full

In [ ]:
!./max_pairwise_product_long_full < input3.txt

40000000000


### Runing multuple test with a bash shell

In [36]:
 %%writefile run_all_test.sh
#!/bin/bash
g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
for i in {1..3}
do
    ./max_pairwise_product_long_full < input$i.txt > ctest$i.txt
    grep -xvFf output$i.txt ctest$i.txt
done

Writing run_all_test.sh


In [37]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

### Test 3  stress tests, 6 testing cases files input4.txt/output4.txt .. input10.txt/output10.txt

---



In [38]:
%%writefile input4.txt
9
91 7 73 29 86 38 22 25 48

Writing input4.txt


In [39]:
%%writefile output4.txt
70815342

Writing output4.txt


In [40]:
%%writefile input5.txt
17
20 56 83 75 13 46 82  5 95 23 17 70 55 47 42 31 16

Writing input5.txt


In [41]:
%%writefile output5.txt
1943733

Writing output5.txt


In [42]:
%%writefile input6.txt
15
100 14 4 37 84 77 54 24 67 3 61 78 12 32 43

Writing input6.txt


In [43]:
%%writefile output6.txt
400910

Writing output6.txt


In [44]:
%%writefile input7.txt
6
99 64 49 65 98 59

Writing input7.txt


In [45]:
%%writefile output7.txt
7498347

Writing output7.txt


In [46]:
%%writefile input8.txt
7
80 79 33 81 9 74 53

Writing input8.txt


In [47]:
%%writefile output8.txt
1107890

Writing output8.txt


In [48]:
%%writefile input9.txt
4
7100 89000 360000 5200000

Writing input9.txt


In [49]:
%%writefile output9.txt
8429255

Writing output9.txt


In [50]:
%%writefile input10.txt
7
88 72 2 40 96 45 57

Writing input10.txt


In [51]:
%%writefile output10.txt
9731141

Writing output10.txt


In [69]:
%%writefile run_all_test.sh
#!/bin/bash
g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full || exit  1

for i in {4..10}
do
    ./max_pairwise_product_long_full <inputi.txt> ctesti.text
    if grep -xvFf outputi.txt ctexti.txt; then
      echo "FAILED"
    else
      echo "OK"
    fi
done

Overwriting run_all_test.sh


In [70]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

./run_all_test.sh: line 6: inputi.txt: No such file or directory
grep: outputi.txt: No such file or directory
OK
./run_all_test.sh: line 6: inputi.txt: No such file or directory
grep: outputi.txt: No such file or directory
OK
./run_all_test.sh: line 6: inputi.txt: No such file or directory
grep: outputi.txt: No such file or directory
OK
./run_all_test.sh: line 6: inputi.txt: No such file or directory
grep: outputi.txt: No such file or directory
OK
./run_all_test.sh: line 6: inputi.txt: No such file or directory
grep: outputi.txt: No such file or directory
OK
./run_all_test.sh: line 6: inputi.txt: No such file or directory
grep: outputi.txt: No such file or directory
OK
./run_all_test.sh: line 6: inputi.txt: No such file or directory
grep: outputi.txt: No such file or directory
OK


### Test 4  time test  measure the execution time running with the the largest worst cases input11.txt  (200000 increasing numbers) and input12.txt (199998 increasing numbers and  200000, 200000 at the end)

In [58]:
# input11.txt 200000 increasing numbers:  1 2 .. 200000

%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input11.txt");

 MyFile << 200000 <<"\n";  // n
 for (int i=1; i <200001; i++){
     MyFile << i <<" ";
 }

  // Close the file
  MyFile.close();
}

Writing WriteTest.cpp


In [71]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [72]:
199999 * 200000

39999800000

In [73]:
%%writefile output11.txt
39999800000

Writing output11.txt


In [74]:
# input12.txt 200000 decreasing numbers and 200000, 199999 ...  2 1

%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input12.txt");

 MyFile << 200000 <<"\n";  // n
 for (int i=1; i <200001; i++){
     MyFile << 200001-i <<" ";
 }

  // Close the file
  MyFile.close();
}

Overwriting WriteTest.cpp


In [75]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [76]:
%%writefile output12.txt
39999800000

Writing output12.txt


### Measure time

https://askubuntu.com/questions/53444/how-can-i-measure-the-execution-time-of-a-terminal-process

In [78]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
!time ./max_pairwise_product_long_full< input11.txt

456185632
^C


In [79]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
!time ./max_pairwise_product_long_full< input12.txt

1562179845
^C


### Calculate the number or and upper bound to the number of instructions executed by the algorithm as a function of n, add in each line a comment with the number of instructions that are executed in that line.

In [80]:
%%writefile max_pairwise_product_long_full.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {              //INSTRUCTIONS executed by line
    long long max_product = 0LL;                                             //1
	  long long aux_product = 0LL;                                             //1
    int n = numbers.size();                                                  //1

    for (int first = 0; first < n; ++first) {                                //2+2n
        for (int second = 0; second < n; ++second) {                         //2n + 2n**2
          if ( first != second ){                                            //n**2
                aux_product = numbers[first] * numbers[second];              //n**2 -n
                if (aux_product > max_product)                               //n**2 -n
                    max_product = aux_product;                               //[0, n**2/2 - n/2] upper bound
          }
        }
    }

                                                                             // --------------
    return max_product;                                                      // 5 + 2*n + [5*n**2, 11*n**2/2  - n/2]  upper bound
}

int main() {                                              //INSTRUCTIONS executed by line
    int n;                                                //1
    std::cin >> n;                                        //1
    std::vector<int> numbers(n);                          //n
    for (int i = 0; i < n; ++i) {                         //2 + 2n
        std::cin >> numbers[i];                           //n
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";     //1 + 1  + 5 + 2*n + [5*n**2, 11*n**2/2  - n/2]

    return 0;                                             // --------------
}                                                         //  11 + 6*n + [5*n**2, 11*n**2/2  - n/2]  upper bound


Overwriting max_pairwise_product_long_full.cpp


### Count executed intructions in code

In [81]:
%%writefile max_pairwise_product_long_full_execints.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers, long long &execints) {      //INSTRUCTIONS executed by line

    long long max_product = 0LL;                                                          //1
    execints++;

	  long long aux_product = 0LL;                                                          //1
    execints++;

    int n = numbers.size();                                                               //1
    execints++;

    execints++;
    for (int first = 0; first < n; ++first) {                                             //2+2n
        execints += 2;

        execints++;
        for (int second = 0; second < n; ++second) {                                       //2n + 2n**2
          execints += 2;

          execints++;
          if ( first != second ){                                                         //n**2
                aux_product = numbers[first] * numbers[second];                           //n**2 -n
                execints++;

                execints++;
                if (aux_product > max_product){                                           //n**2 -n
                    max_product = aux_product;                                            //[0,n**2/2 - n/2)  upper bound
                    execints++;

                }
        }

      }
      execints++;
    }
    execints++;
                                                                                        // --------------
    return max_product;                                                                 // 5 + 2*n + [5*n**2, 11*n**2/2  - n/2] upper bound
}

int main() {                                                 //INSTRUCTIONS executed by line

    long long execints =0LL;

    int n;                                                   //1
    execints++;

    std::cin >> n;                                           //1
    execints++;

    std::vector<int> numbers(n);                             //n
    execints += n;

    execints++;
    for (int i = 0; i < n; ++i) {                            //2 + 2n
        execints += 2;

        std::cin >> numbers[i];
        execints++;                                          //n

    }


    //std::cout << MaxPairwiseProduct(numbers) << "\n";      //1 + 1  + 5 + 2*n + [5*n**2, 11*n**2/2  - n/2] upper bound

    execints++;
    int M = MaxPairwiseProduct(numbers,execints);

    execints++;
    std::cout <<  execints << "\n";
    return 0;                                                // --------------
}                                                            // 11 + 6*n + [5*n**2, 11*n**2/2  - n/2]

Writing max_pairwise_product_long_full_execints.cpp


The number of instructions executed by the algooirithm with a list of size $n$ is  

$$T (n) = 11 + 6n + \left[5n^2, 11\frac{n^2}{2}  - \frac{n}{2}\right)  >  6n^2 = O(n^2)$$

It depends on the size of the lists and  the order of the list.

For the input0.txt
```
10
0 0 0 0 0 0 0 0 0 0
```
$$T = 11 + 6(10) + 5(10)^2   = 561 $$

$$T(10) \in [571, 621)  $$


In [82]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input0.txt

570


For the input1.txt
```
10
1 2 3 4 5 6 7 8 9 10
```
$$T = 11 + 6(10) + 5(10)^2 +  9 + 5 + 4 + 3 + 2 + 2 + 2 + 2 + 1 + 0  = 601 $$

$$T(10) \in [572, 622)  $$


In [83]:
#The number of times that  max_product = aux_product; in the worst case is less than (n**2 -n)/2 -  upper bound

#     1  2  3  4  5  6  7  8  9  10
#  1  x  2  3  4  5  6  7  8  9  10         9
#  2  2  x  6  8  10 12 14 16 18 20         5 14
#  3  3  6  x  12 15 18 21 24 27 30         4 18
#  4  4  8  12 x  20 24 28 32 36 40         3 21
#  5  5  10 15 20 x  30 35 40 45 50         2 23
#  6  6  12 18 24 30 x  42 48 54 60         2 25
#  7  7  14 21 28 35 42 x  56 63 70         2 27
#  8  8  18 27 36 45 54 63 72 x  90         1 30
#  10 10 20 30 40 50 60 70 60 90 x          0

In [84]:
11 + 6*(10) + 5*(10)*(10) +  9 + 5 + 4 + 3 + 2 + 2 + 2 + 2 + 1 + 0

601

In [85]:
n = 10
print( "[", 11 + 6*n + 5*n**2,  ",", 11 + 6*n  + 5.5*n**2, ")")

[ 571 , 621.0 )


In [86]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input1.txt

600



For the input2.txt
```
10
10 9 8 7 6 5 4 3 2 1
```
$$T = 11 + 6*(10) + 5*(10)**2 +  1  = 571 $$

In [87]:
11 + 6*(10) + 5*(10)**2 +  1

572

In [88]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input2.txt

571


Calculate estimated time by instruction and number of instructions by second

For the input11.txt
```
200000
1 2 ..  199999 2000
```
$$T = 11 + 6*(200000) + 5.5*(200000)**2  = 200001200012  $$

In [91]:
t = 11 + 6*(200000)+ 5.5*(200000)**2
print(t)

220001200011.0


In [92]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input11.txt

1
1
2
^C


In [93]:
# Estimated time by instruction
# use the real reported  !time ./max_pairwise_product_long_full< input11.txt in seconds intead of the 295

295/200003177092


1.4749765693187072e-09

In [94]:
# Estimated number of instructions by second

200003177092/295

677976871.4983051

### Estimate and verify the maximun n that can be solved in 1 second

$$(5.5n^2 + 5.5n + 11) \times 1.47*10^{-9} s = 1 s$$

$$ n \approx 11000$$


In [95]:
# Verify input13.txt
%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input13.txt");

 MyFile << 11000 <<"\n";  // n
 for (int i=1; i <11001; i++){
     MyFile << i <<" ";
 }

  // Close the file
  MyFile.close();
}

Overwriting WriteTest.cpp


In [96]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [97]:
10999 * 11000

120989000

In [98]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
!time ./max_pairwise_product_long_full< input13.txt

120989000

real	0m0.855s
user	0m0.847s
sys	0m0.004s


## Python

In [99]:
%%writefile max_pairwise_product.py

def max_pairwise_product(numbers):
  max = 0
  aux = 0
  n = len(numbers)
  for first in range(n):
    for second in range(first + 1, n):
      aux = numbers[first] * numbers[second]
      if aux > max:
        max = aux
  return max
n = int(input())
numbers = [int(x) for x in input().split()]
print(max_pairwise_product(numbers))

Writing max_pairwise_product.py


In [102]:
# verify
!time python3 max_pairwise_product.py < input13.txt

120989000

real	0m9.443s
user	0m8.920s
sys	0m0.015s


# For the follwing solutions rquired  the same tests and analysis performed for the first solution

## Solutions in C++ and Python that uses
$$\frac{n^2 - n}{2} = O(n^2)\ \ \text{products}$$

In [103]:
%%writefile max_pairwise_product.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int MaxPairwiseProduct(const std::vector<int>& numbers) {
    int max_product = 0;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = first + 1; second < n; ++second) {
            max_product = std::max(max_product,
                numbers[first] * numbers[second]);
        }
    }

    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}

Writing max_pairwise_product.cpp


In [104]:
%%writefile max_pairwise_product_long.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {              //INSTRUCTIONS executed by line
    long long max_product = 0LL;                                             //1
	  long long aux_product = 0LL;                                             //1
    int n = numbers.size();                                                  //1

    for (int first = 0; first < n; ++first) {                                //2+2n
        for (int second = first + 1; second < n; ++second) {                 //2n + (n-1)n
			       aux_product =  (long long) numbers[first] * numbers[second];    //(n-1)n/2
            if (max_product < aux_product)  {                                //(n-1)n/2
				        max_product = aux_product;                                   //[1,(n-1)n/2)  upper bound
             }
        }
    }
                                                                             // --------------
    return max_product;                                                      // 5 + 4n + [2(n-1)n+1, 2.5 (n-1)n) upper bound
}

int main() {                                              //INSTRUCTIONS executed by line
    int n;                                                //1
    std::cin >> n;                                        //1
    std::vector<int> numbers(n);                          //n
    for (int i = 0; i < n; ++i) {                         //2 + 2n
        std::cin >> numbers[i];                           //n
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";     //1 + 1  + 5 + 4n + [2(n-1)n+1, 2.5 (n-1)n) upper bound
    return 0;                                             // --------------
}                                                         // 11 + 7n + [2(n-1)n+1, 2.5 (n-1)n) upper bound


Writing max_pairwise_product_long.cpp


In [105]:
!g++ max_pairwise_product_long.cpp -o max_pairwise_product_long
!time ./max_pairwise_product_long < input13.txt

120989000

real	0m0.425s
user	0m0.418s
sys	0m0.001s


In [106]:
!g++ max_pairwise_product.cpp -o max_pairwise_product
!time ./max_pairwise_product < input13.txt

120989000

real	0m0.552s
user	0m0.549s
sys	0m0.002s


# Sort solution in C++ and Python

$$ O(n \lg n)$$

**Which sorting algorithm is used in GCC?**
https://stackoverflow.com/questions/7221257/which-sorting-algorithm-is-used-in-gcc

* ...  the algorithm is called introsort, a combination between Insertionsort, Quicksort & Heapsort. ...

**Introsort**
https://en.wikipedia.org/wiki/Introsort

Introsort or introspective sort is a hybrid sorting algorithm that provides both fast average performance and (asymptotically) optimal worst-case performance. It begins with quicksort, it switches to heapsort when the recursion depth exceeds a level based on (the logarithm of) the number of elements being sorted and it switches to insertion sort when the number of elements is below some threshold

*   Class	Sorting algorithm
*   Data structure	Array
*   Worst-case performance	$O(n \lg n)$
*   Average performance	$O(n \lg n)$

**Timsort**
https://en.wikipedia.org/wiki/Timsort

Timsort is a hybrid, stable sorting algorithm, derived from merge sort and insertion sort, designed to perform well on many kinds of real-world data. It was implemented by Tim Peters in 2002 for use in the Python programming language.

*   Class	Sorting algorithm
*   Data structure	Array
*   Worst-case performance	$O(n \lg n)$
*   Best-case performance $O(n)$
*   Average performance	$O(n \lg n)$
*   Worst-case space complexity $O(n)$

In [107]:
%%writefile max_pairwise_product_with_sort.py
#Python3
if __name__ == "__main__":
    n = int(input())
    lista = list(map(int, input().split()))
    lista.sort()
    print(lista[n - 1] * lista[n - 2])

Writing max_pairwise_product_with_sort.py


Test with inputs 1 and 2



In [128]:
%%writefile MaxPairwiseProductWithSort.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProductWithSort(std::vector<int>& numbers) {
    int n = numbers.size();
    std::sort(numbers.begin(), numbers.end());
    return (long long)numbers[n - 1] * numbers[n - 2];
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProductWithSort(numbers) << "\n";
    return 0;
}

Writing MaxPairwiseProductWithSort.cpp


In [129]:
!g++ MaxPairwiseProductWithSort.cpp -o MaxPairwiseProductWithSort

In [130]:
!./MaxPairwiseProductWithSort < input1.txt
!./MaxPairwiseProductWithSort < input2.txt

90
90


In [132]:
!python max_pairwise_product_with_sort.py < input1.txt
!python max_pairwise_product_with_sort.py < input2.txt

90
90


Test and time it with input 3

In [133]:
!time ./MaxPairwiseProductWithSort < input3.txt

40000000000

real	0m0.004s
user	0m0.001s
sys	0m0.002s


In [136]:
%%time
!python max_pairwise_product_with_sort.py < input3.txt

40000000000
CPU times: user 5.6 ms, sys: 1.52 ms, total: 7.12 ms
Wall time: 106 ms


20 stress tests

In [145]:
%%writefile stress_tests.py
import random

if __name__ == "__main__":
    for i in range(20):
        n = random.randint(2, 19)
        lista = [random.randint(17, 25631) for i in range(n)]
        lista.sort()
        print(f"n: {n} Product: {lista[n - 1] * lista[n - 2]}")




Overwriting stress_tests.py


In [146]:
!python stress_tests.py

n: 7 Product: 548310113
n: 12 Product: 533446046
n: 11 Product: 167874107
n: 12 Product: 463226982
n: 8 Product: 557450111
n: 14 Product: 597609337
n: 19 Product: 540017298
n: 13 Product: 576302841
n: 10 Product: 567275420
n: 14 Product: 610231193
n: 11 Product: 596345520
n: 13 Product: 637484120
n: 7 Product: 497504490
n: 4 Product: 271963769
n: 6 Product: 500399592
n: 13 Product: 556437546
n: 10 Product: 542533930
n: 12 Product: 499093056
n: 15 Product: 581365210
n: 3 Product: 112538070


In [148]:
%%writefile stress_tests_C.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <cstdlib> // For rand() and srand()
#include <ctime>   // For time()


long long MaxPairwiseProductWithSort(std::vector<int>& numbers) {
    int n = numbers.size();
    std::sort(numbers.begin(), numbers.end());
    return (long long)numbers[n - 1] * numbers[n - 2];
}

int main() {
    srand(time(0)); // Seed the random number generator

    for (int i = 0; i < 20; ++i) {
        int n = rand() % 10 + 2; // Generate a random number of elements (between 2 and 11)
        std::vector<int> numbers(n);
        for (int j = 0; j < n; ++j) {
            numbers[j] = rand() % 1000; // Generate random numbers between 0 and 999
        }

        long long result = MaxPairwiseProductWithSort(numbers);
        std::cout << "Test case " << i + 1 << ": n = " << n << ", Max Pairwise Product = " << result << std::endl;
    }

    return 0;
}

Writing stress_tests_C.cpp


In [149]:
!g++ stress_tests_C.cpp -o stress_tests_C
!./stress_tests_C

Test case 1: n = 5, Max Pairwise Product = 407037
Test case 2: n = 6, Max Pairwise Product = 736470
Test case 3: n = 9, Max Pairwise Product = 763014
Test case 4: n = 3, Max Pairwise Product = 605150
Test case 5: n = 9, Max Pairwise Product = 536406
Test case 6: n = 7, Max Pairwise Product = 717184
Test case 7: n = 8, Max Pairwise Product = 756243
Test case 8: n = 6, Max Pairwise Product = 790152
Test case 9: n = 8, Max Pairwise Product = 858452
Test case 10: n = 5, Max Pairwise Product = 746982
Test case 11: n = 4, Max Pairwise Product = 733516
Test case 12: n = 11, Max Pairwise Product = 751640
Test case 13: n = 3, Max Pairwise Product = 381000
Test case 14: n = 8, Max Pairwise Product = 740230
Test case 15: n = 5, Max Pairwise Product = 321855
Test case 16: n = 3, Max Pairwise Product = 527120
Test case 17: n = 4, Max Pairwise Product = 382782
Test case 18: n = 5, Max Pairwise Product = 439817
Test case 19: n = 10, Max Pairwise Product = 969184
Test case 20: n = 7, Max Pairwise Prod

Count executed instructions in code and verify with inputs 1,2 and 3


In [154]:
%%writefile count_operations.py
import math

if __name__ == "__main__":
    inst=0
    n = int(input())
    inst+=1
    lista = list(map(int, input().split()))
    inst+= 4*n
    lista.sort()
    inst+=  n * math.log2(n)
    print(lista[n - 1] * lista[n - 2])
    inst+=1
    print(f"Instrucciones ejecutadas: {int(inst)}\n")

Overwriting count_operations.py


In [155]:
!python count_operations.py < input1.txt
!python count_operations.py < input2.txt
!python count_operations.py < input3.txt

90
Instrucciones ejecutadas: 75

90
Instrucciones ejecutadas: 75

40000000000
Instrucciones ejecutadas: 41



In [162]:
%%writefile count_operations_C.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include

long long MaxPairwiseProductWithSort(std::vector<int>& numbers, long long& execints) {
    int n = numbers.size();
    execints++;
    std::sort(numbers.begin(), numbers.end());
    execints += n * log2(n); // Approximate number of operations for sort
    execints += 2; // Accessing elements
    return (long long)numbers[n - 1] * numbers[n - 2];
}

int main() {
    long long execints = 0;
    int n;
    std::cin >> n;
    execints++;
    std::vector<int> numbers(n);
    execints += n;
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
        execints += 2;
    }

    long long result = MaxPairwiseProductWithSort(numbers, execints);
    std::cout << result << "\n";
    std::cout << "Number of instructions: " << execints << std::endl;
    return 0;
}

Overwriting count_operations_C.cpp


In [163]:
!g++ count_operations_C.cpp -o count_operations_C

In [166]:
!./count_operations_C < input1.txt
!./count_operations_C < input2.txt
!./count_operations_C < input3.txt

90
Number of instructions: 67
90
Number of instructions: 67
40000000000
Number of instructions: 37


Calculate estimated time by instruction and number of instructions by second

In [168]:


%%writefile max_pairwise_product_with_sort_and_instructions_with_time_and_IbS.py
#Python3
import math
import time
if __name__ == "__main__":
    inicio = time.time()
    inst=0
    n = int(input())
    inst+=1
    lista = list(map(int, input().split()))
    inst+= 4*n
    lista.sort()
    inst+=  n * math.log2(n)
    print(lista[n - 1] * lista[n - 2])
    inst+=1
    print(f"Instrucciones ejecutadas: {int(inst)}")
    fin = time.time()
    tiempo_total = fin - inicio
    instructionTime = tiempo_total/inst
    instructionBySecond= inst/tiempo_total

    print(f"Tiempo total de ejecución: {tiempo_total*1000} milisegundos")
    print(f"Tiempo estimado de cada instruccion: {instructionTime*1000} milisegundos")
    print(f"Instrucciones por segundo: {int(instructionBySecond)} ")



Writing max_pairwise_product_with_sort_and_instructions_with_time_and_IbS.py


In [169]:
!python max_pairwise_product_with_sort_and_instructions_with_time_and_IbS.py < input13.txt

120989000
Instrucciones ejecutadas: 191679
Tiempo total de ejecución: 2.4623870849609375 milisegundos
Tiempo estimado de cada instruccion: 1.2846385198091028e-05 milisegundos
Instrucciones por segundo: 77842909 


### Estimate and verify the maximun n that can be solved in 1 second

In [182]:
import time
import random
n = 0 #numero de instrucciones

def instruction(numbers):
    m = len(numbers)
    biggest_1 = float('-inf')
    biggest_2 = float('-inf')
    global n
    n += 3

    for i in range(m):
        new = numbers[i]
        temp = biggest_1
        n += 2
        if new > biggest_1:
            biggest_2 = biggest_1
            biggest_1 = new
            n += 2
        if temp > biggest_2:
            biggest_2 = temp
            n += 1
        n += 2
    return n
# Ejecutar la función y medir el tiempo de ejecución
start_time = time.time()
numbers = [random.randint(1, 100) for _ in range(10)]
instruction = instruction(numbers) # Suponiendo que 'numbers' es una lista válida
end_time = time.time()

# Calcular el número de instrucciones por segundo
elapsed_time = end_time - start_time
instructions_per_second = instruction / elapsed_time
#print(start_time)
#print(end_time)
print(f"Número de instrucciones por segundo: {instructions_per_second}")

Número de instrucciones por segundo: 189769.9870729455


## Efficient solution with largest and second-largest in an list, array or vector in C++ and Python

 $$T(n) < 10 n  = O(n)$$





Add in each line a coment with the number of intstrcution that are executed in that line

In [185]:
%%writefile efficientSolution.py

def efficientSolution(numbers):  # 1 instrucción para definir la función
    largest = 0  # 1 instrucción para asignar el valor 0 a largest
    secondLargest = 0  # 1 instrucción para asignar el valor 0 a secondLargest
    for i in numbers:  # 1 instrucción para iniciar el bucle
        if i > largest:  # 1 instrucción para la comparación i > largest
            secondLargest = largest  # 1 instrucción para asignar largest a secondLargest
            largest = i  # 1 instrucción para asignar i a largest
        elif i > secondLargest:  # 1 instrucción para la comparación i > secondLargest
            secondLargest = i  # 1 instrucción para asignar i a secondLargest

    result = secondLargest * largest  # 1 instrucción para la multiplicación y asignación a result
    return result  # 1 instrucción para retornar el resultado

n = int(input())  # 1 instrucción para leer la entrada y convertirla a entero
numbers = [int(i) for i in input().split()]  # 2 instrucciones: leer entrada y convertir a lista de enteros
print(efficientSolution(numbers))  # 1 instrucción para llamar a la función y mostrar el resultado

Writing efficientSolution.py


Test with inputs 1 and 2



In [186]:
!python3 efficientSolution.py < input1.txt
!python3 efficientSolution.py < input2.txt

90
90


Test and time it with input 3

In [187]:
!time python3 efficientSolution.py < input3.txt

40000000000

real	0m0.148s
user	0m0.039s
sys	0m0.012s


20 stress tests

In [197]:
# prompt: write a function that generates 20 stress tests for the efficienSolution.py file

import random

def generate_stress_test(n):
    numbers = [random.randint(-2**32,2**32) for _ in range(n)]
    return numbers

def main():
    for i in range(20):
        n = random.randint(2, 2**22)  # Vary the input size
        numbers = generate_stress_test(n)

        # Construct the input string
        input_str = f"{n}\n"
        input_str += " ".join(map(str, numbers)) + "\n"

        # Write the input to a file (optional, for debugging)
        #with open(f"stress_test_{i}.txt", "w") as f:
        #    f.write(input_str)

        # Run the efficientSolution.py script with this input
        print(f"Running stress test {i + 1} with n = {n}")
        # Uncomment these lines to run efficientSolution.py with each stress test
        # Use a shell command to redirect the input string
        #!python efficientSolution.py <<< "$input_str"

if __name__ == "__main__":
    main()

Running stress test 1 with n = 1788651
Running stress test 2 with n = 3607814
Running stress test 3 with n = 3276867
Running stress test 4 with n = 263635
Running stress test 5 with n = 777592
Running stress test 6 with n = 2249349
Running stress test 7 with n = 159428
Running stress test 8 with n = 401487
Running stress test 9 with n = 1335682
Running stress test 10 with n = 1715743
Running stress test 11 with n = 2122237
Running stress test 12 with n = 4133795
Running stress test 13 with n = 3428032
Running stress test 14 with n = 652350
Running stress test 15 with n = 1598675
Running stress test 16 with n = 3172278
Running stress test 17 with n = 1428782
Running stress test 18 with n = 1254144
Running stress test 19 with n = 1826434
Running stress test 20 with n = 3148109


Count executed instructions in code and verify with inputs 1,2 and 3


In [198]:

%%writefile efficientSolutionexcec.py
exceintspy= 0
def efficientSolution(numbers):
    global exceintspy
    largest = 0
    exceintspy+=1
    secondLargest = 0
    exceintspy+=1
    exceintspy+=1
    for i in numbers:
        exceintspy+=2
        exceintspy+=1
        if i > largest:
            secondLargest = largest
            exceintspy+=1
            largest = i
            exceintspy+=1
        elif i > secondLargest:
            secondLargest = i
            exceintspy+=1
    exceintspy+=1
    result= secondLargest * largest
    exceintspy+=2 # multiplciation instruction
    return result

n = int(input())
exceintspy+=1
numbers = [int(i) for i in input().split()]
exceintspy+=n
print(efficientSolution(numbers))
exceintspy+=1
print(f"Execution Times: {exceintspy}")

Writing efficientSolutionexcec.py


In [199]:
%%writefile run_all_test.sh
for i in {1..3}
do
    echo "Input #$i"
    python3 efficientSolutionexcec.py < input$i.txt
done

Overwriting run_all_test.sh


In [200]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

Input #1
90
Execution Times: 68
Input #2
90
Execution Times: 51
Input #3
40000000000
Execution Times: 40


Calculate estimated time by instruction and number of instructions by second

In [201]:
!time python3 efficientSolutionexcec.py< input11.txt > ptest.txt
!grep -xvFf output11.txt ptest.txt


real	0m0.162s
user	0m0.131s
sys	0m0.029s
Execution Times: 1200008


### Estimate and verify the maximun n that can be solved in 1 second

In [202]:


%%writefile efficientSolution.py
def efficientSolution(numbers):
    largest = 0                     #1
    secondLargest = 0               #1
    for i in numbers:               #2+2n
        if i > largest:
            secondLargest = largest
            largest = i
        elif i > secondLargest:
            secondLargest = i
    #WORST CASE 3n
    result= secondLargest * largest    #3
    return result

n = int(input())                            #1
numbers = [int(i) for i in input().split()] #n
print(efficientSolution(numbers))           #1



Overwriting efficientSolution.py


In [203]:
!python3 efficientSolution.py < input13.txt

120989000


## Super efficient solution

Whitout storing the list in memory, only stpring the largest and second largest and whitout calling a function, add in each line a coment with the number of instructions that are executed in that line

in C++ and Python


 $$T(n) < 13 n  = O(n)$$


In [204]:
%%writefile max_efficient.py

def max_pairwise_efficient():
    n = int(input())
    values = input().split()
    biggest_1 = float('-inf')
    biggest_2 = float('-inf')

    for new in map(int, values):
        temp = biggest_1
        if new > biggest_1:
            biggest_2 = biggest_1
            biggest_1 = new
        if temp > biggest_2:
            biggest_2 = temp

    return biggest_1 * biggest_2

if __name__ == '__main__':
    print(max_pairwise_efficient())

Writing max_efficient.py


In [205]:
!python3 max_efficient.py

5
4 5 63 3
3969


Test with inputs 1 and 2




In [206]:
%%writefile max_efficient2.py

def max_pairwise_efficient():
    n = int(input())
    values = input().split()
    biggest_1 = float('-inf')
    biggest_2 = float('-inf')

    for new in map(int, values):
        temp = biggest_1
        if new > biggest_1:
            biggest_2 = biggest_1
            biggest_1 = new
        elif temp > biggest_2:
            biggest_2 = temp

    return biggest_1 * biggest_2

if __name__ == '__main__':
    print(max_pairwise_efficient())

Writing max_efficient2.py


In [207]:
%%writefile run_all_test2.sh
#comparing all the executions to check the differences between the outputs
for i in {1..2}
do
    python3 max_efficient2.py < inputi.txt
done

Writing run_all_test2.sh


In [208]:
!chmod u+x run_all_test2.sh
!./run_all_test.sh

Input #1
90
Execution Times: 68
Input #2
90
Execution Times: 51
Input #3
40000000000
Execution Times: 40


Test and time it with input 3

In [209]:
!time python3 max_efficient2.py < input3.txt > pythonOutput3.txt


real	0m0.035s
user	0m0.023s
sys	0m0.012s


20 stress tests

In [210]:
import random

def max_pairwise_efficient(numbers):
    n = len(numbers)
    biggest_1 = float('-inf')
    biggest_2 = float('-inf')

    for i in range(n):
        new = numbers[i]
        temp = biggest_1
        if new > biggest_1:
          biggest_2 = biggest_1
          biggest_1 = new
        if temp > biggest_2: biggest_2 = temp

    return biggest_1 * biggest_2

# 20 stress tests
for _ in range(20):
    n = random.randint(2, 12)  # Use a large n
    numbers = [random.randint(1, 10**9) for _ in range(n)]  # Generate n random numbers between 1 and 10^9

    print(f"Testing with the list: {numbers}")
    print(f"The efficient solution gives the result {max_pairwise_efficient(numbers)}")


Testing with the list: [785704541, 701924234, 316632882, 105139023, 22856322, 383485081, 254437908]
The efficient solution gives the result 617331625748020681
Testing with the list: [488984158, 552631179, 4969466, 365688678, 32251994, 715134744, 897647723]
The efficient solution gives the result 641939074589787912
Testing with the list: [694942766, 439149915, 241353918, 549203566, 600867881, 694084771, 123684180, 347917114, 648774472]
The efficient solution gives the result 482945448015730756
Testing with the list: [496475578, 13874464, 403356525, 804387868, 446023613, 839160729, 985026062, 614222230, 347024726]
The efficient solution gives the result 970276342819227844
Testing with the list: [541792467, 929672799, 252331012, 903680298, 735293761, 148511958, 614293522, 391245022, 254225523, 814367548, 30782250]
The efficient solution gives the result 864291513200494401
Testing with the list: [5791925, 986555849]
The efficient solution gives the result 5714057485719325
Testing with the 

Count executed instructions in code and verify with inputs 1,2 and 3

In [211]:
%%writefile max_pairwise_efficient.py
import random
execints = 0

def max_pairwise_efficient(numbers):
    global execints
    n = len(numbers)
    biggest_1 = float('-inf')
    biggest_2 = float('-inf')
    execints += 3

    for i in range(n):
        new_number = numbers[i]
        temp = biggest_1
        execints += 2
        if new_number > biggest_1:
            biggest_2 = biggest_1
            biggest_1 = new_number
            execints += 2
        if temp > biggest_2:
            biggest_2 = temp
            execints += 2
        execints += 2

    execints += 1
    return biggest_1 * biggest_2

if __name__ == '__main__':
    n = int(input())
    input_line = input()
    numbers = list(map(int, input_line.split()))  # Divide la cadena en números y convierte cada uno a entero
    max_pairwise_efficient(numbers)
    execints += 1

    print("El numero de instrucciones es:", execints)

Overwriting max_pairwise_efficient.py


In [212]:
%%writefile run_all_test.sh
for i in {1..3}
do
    echo "Input #$i"
    python3 max_pairwise_efficient.py < input$i.txt
done

Overwriting run_all_test.sh


In [214]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

Input #1
El numero de instrucciones es: 65
Input #2
El numero de instrucciones es: 49
Input #3
El numero de instrucciones es: 37


Calculate estimated time by instruction and number of instructions by second

In [215]:
!time python3 max_pairwise_efficient.py < input11.txt > pythonOutputSort11.txt
!python3 max_pairwise_efficient.py < input11.txt
num_inst_per_seg = 1200005/0.165
print(f"{num_inst_per_seg} tiempo por instrucción")


real	0m0.274s
user	0m0.224s
sys	0m0.037s
El numero de instrucciones es: 1200005
7272757.575757575 tiempo por instrucción


## Perform 100 randomized automated tests in C++ and Python with a porgram  compares the answers,for the same 100 random input instances, of the super efficient solution $(T(n) < 10 n = O(n))$ against the sorting solution  $(T(n)  = O(n \lg{n}))$,  if the answers are not the same stops and displays "Error" listing  the input and the two answers; if the 100 tests are completed without stoping displays "Pass".

In [216]:
%%writefile max_pairwise_super_efficient_against_sorting_stress_tests.py
import random

for i in range(100):
  n = random.randint(2, 99)
  numero_M = 0                                # 1
  segundo_Numero_M = 0
  lista = []
  for j in range(n):
    comp=random.randint(17, 25631)
    lista.append(comp)
    if comp > numero_M:                    # n
      segundo_Numero_M = numero_M      # 0 o n
      numero_M = comp                  # 0 o n
    elif comp > segundo_Numero_M:           # n
      segundo_Numero_M = comp             # 0 o n
  lista.sort()
  sol_sorting=lista[n - 1] * lista[n - 2]
  sol_super_e = numero_M * segundo_Numero_M


  if sol_sorting != sol_super_e:
    print("Error")
    break
  else:
      print(f"{i+1}. SUPER EFFICENT SOLITION: n= {n} Product= {numero_M * segundo_Numero_M}")
      print(f"{i+1}. SORTING SOLUTION: n= {n} Product: {lista[n - 1] * lista[n - 2]}")
  if i==99:
    print("Pass")

Writing max_pairwise_super_efficient_against_sorting_stress_tests.py


In [217]:
!python max_pairwise_super_efficient_against_sorting_stress_tests.py

1. SUPER EFFICENT SOLITION: n= 33 Product= 614647572
1. SORTING SOLUTION: n= 33 Product: 614647572
2. SUPER EFFICENT SOLITION: n= 89 Product= 617100939
2. SORTING SOLUTION: n= 89 Product: 617100939
3. SUPER EFFICENT SOLITION: n= 68 Product= 613577430
3. SORTING SOLUTION: n= 68 Product: 613577430
4. SUPER EFFICENT SOLITION: n= 48 Product= 624780979
4. SORTING SOLUTION: n= 48 Product: 624780979
5. SUPER EFFICENT SOLITION: n= 92 Product= 631660134
5. SORTING SOLUTION: n= 92 Product: 631660134
6. SUPER EFFICENT SOLITION: n= 33 Product= 552767919
6. SORTING SOLUTION: n= 33 Product: 552767919
7. SUPER EFFICENT SOLITION: n= 94 Product= 646837473
7. SORTING SOLUTION: n= 94 Product: 646837473
8. SUPER EFFICENT SOLITION: n= 45 Product= 578644440
8. SORTING SOLUTION: n= 45 Product: 578644440
9. SUPER EFFICENT SOLITION: n= 19 Product= 487836480
9. SORTING SOLUTION: n= 19 Product: 487836480
10. SUPER EFFICENT SOLITION: n= 93 Product= 651601782
10. SORTING SOLUTION: n= 93 Product: 651601782
11. SUPE